internal/engine/experiment/quicping/quicping.go

```diff
@@ -0,0 +1,443 @@
+			sess.Logger().Infof(fmt.Sprintf("response dissection failed: %s", err))
+			continue
+		}
+		// propagate receive information
+		out <- responseInfo{raw: resp, t: respTime, dstID: hex.EncodeToString(dst), versions: supportedVersions}
+
+		sess.Logger().Infof("PING got response from %s", addr)
+	}
+}
+
+// Run implements ExperimentMeasurer.Run.
+func (m *Measurer) Run(
+	ctx context.Context,
+	sess model.ExperimentSession,
+	measurement *model.Measurement,
+	callbacks model.ExperimentCallbacks,
+) error {
+	host := string(measurement.Input)
+	// allow URL input
+	if u, err := url.ParseRequestURI(host); err == nil {
+		host = u.Host
+	}
+	service := net.JoinHostPort(host, m.config.port())
+	udpAddr, err := net.ResolveUDPAddr("udp4", service)
+	if err != nil {
+		return err
+	}
+	rep := m.config.repetitions()
+	tk := &TestKeys{
+		Domain:      host,
+		Repetitions: rep,
+	}
+	measurement.TestKeys = tk
+
+	// create UDP socket
+	pconn, err := m.config.networkLibrary().ListenUDP("udp", &net.UDPAddr{})
+	if err != nil {
+		return err
+	}
+	defer pconn.Close()
+
+	// set context and read timeouts
+	deadline := time.Duration(rep*5) * time.Second
+	pconn.SetReadDeadline(time.Now().Add(deadline))
+	ctx, cancel := context.WithTimeout(ctx, time.Second*time.Duration(5*rep))
+	defer cancel()
+
+	sendInfoChan := make(chan requestInfo)
+	recvInfoChan := make(chan responseInfo)
+	sendMap := make(map[string]*requestInfo)
+
+	// start sender and receiver goroutines
+	go m.sender(ctx, pconn, udpAddr, sendInfoChan, sess)
+	go m.receiver(ctx, pconn, recvInfoChan, sess)
+L:
+	for {
+		select {
+		case req := <-sendInfoChan: // a new ping was sent
+			if req.err != nil {
+				tk.Pings = append(tk.Pings, &SinglePing{
+					ConnIdDst:   req.dstID,
+					ConnIdSrc:   req.srcID,
+					Failure:     archival.NewFailure(req.err),
+					Request:     &model.ArchivalMaybeBinaryData{Value: string(req.raw)},
+					RequestTime: formatTime(req.t),
+				})
+				continue
+			}
+			sendMap[req.srcID] = &req
+
+		case resp := <-recvInfoChan: // a new response has been received
+			var (
+				req *requestInfo
+				ok  bool
+			)
+			// match response to request
+			if req, ok = sendMap[resp.dstID]; !ok {
+				continue // ignore any version negotiation response with an unknown destination ID
+			}
+			delete(sendMap, resp.dstID)
+			tk.Pings = append(tk.Pings, &SinglePing{
+				ConnIdDst:         req.dstID,
+				ConnIdSrc:         req.srcID,
+				Failure:           archival.NewFailure(resp.err),
+				Request:           &model.ArchivalMaybeBinaryData{Value: string(req.raw)},
+				RequestTime:       formatTime(req.t),
+				Response:          &model.ArchivalMaybeBinaryData{Value: string(resp.raw)},
+				ResponseTime:      formatTime(resp.t),
+				SupportedVersions: resp.versions,
+			})
+			if len(tk.Pings) >= int(rep) {
+				break L // break when we collected all ping responses
+			}
+
+		case <-ctx.Done():
+			break L
+
+		default:
```

Comment:
```suggestion
```

Having a default arm here strikes me as incorrect. It means you are not blocked in either of the two arms of the select and instead you are busy looping. I think you should double check the CPU utilization during the experiment and then remove this default arm of the select statement.